worker/src/main/java/gov/cms/ab2d/worker/processor/CoverageMappingCallable.java

```diff
@@ -0,0 +1,296 @@
+        this.skipBillablePeriodCheck = skipBillablePeriodCheck;
+    }
+
+    public boolean isCompleted() {
+        return completed.get();
+    }
+
+    public CoverageMapping getCoverageMapping() {
+        return coverageMapping;
+    }
+
+    @Override
+    public CoverageMapping call() {
+
+        int month = coverageMapping.getPeriod().getMonth();
+        String contractNumber = coverageMapping.getContract().getContractNumber();
+
+        final Set<Identifiers> patientIds = new HashSet<>();
+        int bundleNo = 1;
+        try {
+
+
+            log.info("retrieving contract membership for Contract {}-{}-{} bundle #{}",
+                    contractNumber, year, month, bundleNo);
+
+            BFDClient.BFD_BULK_JOB_ID.set(coverageMapping.getJobId());
+
+            try {
+                Bundle bundle = getBundle(contractNumber, month);
+                patientIds.addAll(extractAndFilter(bundle));
+
+                String availableLinks = bundle.getLink().stream()
+                        .map(link -> link.getRelation())
+                        .collect(joining(" , "));
+                log.info("retrieving contract membership for Contract {}-{}-{} bundle #{}, available links {}",
+                        contractNumber, year, month, bundleNo, availableLinks);
+
+                if (bundle.getLink(Bundle.LINK_NEXT) == null) {
+                    log.info("retrieving contract membership for Contract {}-{}-{} bundle #{}, does not have a next link",
+                            contractNumber, year, month, bundleNo);
+                }
+
+                while (bundle.getLink(Bundle.LINK_NEXT) != null) {
+
+                    bundleNo += 1;
+
+                    log.info("retrieving contract membership for Contract {}-{}-{} bundle #{}",
+                            contractNumber, year, month, bundleNo);
+
+                    bundle = bfdClient.requestNextBundleFromServer(bundle);
+
+                    availableLinks = bundle.getLink().stream()
+                            .map(link -> link.getRelation() + " -> " + link.getUrl())
+                            .collect(joining(" , "));
+
+                    log.info("retrieving contract membership for Contract {}-{}-{} bundle #{}, available links {}",
+                            contractNumber, year, month, bundleNo, availableLinks);
+
+                    if (bundle.getLink(Bundle.LINK_NEXT) == null) {
+                        log.info("retrieving contract membership for Contract {}-{}-{} bundle #{}, does not have a next link",
+                                contractNumber, year, month, bundleNo);
+                    }
+
+                    patientIds.addAll(extractAndFilter(bundle));
+                }
+            } catch (InternalErrorException ie) {
+                // Catch edge case bug where (number patients) mod (bundle size) == 0
+                // Extra bundle link returned that has no data in it which causes exception
```

Comment:
This should be fixed by the changes made by me, Christopher and Karl.  Please re-verify and remove if no longer needed.

The same goes for PatientContractCallable.  In fact, there may be an opportunity to create an abstract base class for these two classes and do a pull up refactoring of common functionality.

Reply:
I'll check whether the fix works. As for the rest the goal is to remove the PatientContractCallable completely so I'm not sure refactoring is a great idea.

Reply:
The fix does work